BootloaderCommonPkg/Library/ConsoleOutLib/ConsoleOutLib.c

```diff
@@ -0,0 +1,218 @@
+/** @file
+  Provide Console output library functions.
+
+  Copyright (c) 2018, Intel Corporation. All rights reserved.<BR>
+  This program and the accompanying materials
+  are licensed and made available under the terms and conditions of the BSD License
+  which accompanies this distribution.  The full text of the license may be found at
+  http://opensource.org/licenses/bsd-license.php.
+
+  THE PROGRAM IS DISTRIBUTED UNDER THE BSD LICENSE ON AN "AS IS" BASIS,
+  WITHOUT WARRANTIES OR REPRESENTATIONS OF ANY KIND, EITHER EXPRESS OR IMPLIED.
+
+**/
+
+#include <PiPei.h>
+#include <Library/BaseLib.h>
+#include <Library/BaseMemoryLib.h>
+#include <Library/BootloaderCommonLib.h>
+#include <Library/DebugLogBufferLib.h>
+#include <Guid/LoaderPlatformDataGuid.h>
+#include <Library/GraphicsLib.h>
+#include <Library/PrintLib.h>
+#include <Library/SerialPortLib.h>
+#include <Library/HobLib.h>
+#include <Library/ConsoleOutLib.h>
+
+/**
+  Initialize the virtual console framebuffer back-end.
+
+  @retval EFI_UNSUPPORTED Could not initilaize the console.
+  @retval EFI_SUCCESS     Console initialized.
+**/
+EFI_STATUS
+EFIAPI
+ConsoleInitFrameBuffer (
+  VOID
+  )
+{
+  FRAME_BUFFER_INFO       *FrameBuffer;
+  FRAME_BUFFER_CONSOLE    *Console;
+  DEBUG_LOG_BUFFER_HEADER *LogBufHdr;
+
+  FrameBuffer = (FRAME_BUFFER_INFO *)GetGuidHobData (NULL, NULL, &gLoaderFrameBufferInfoGuid);
+  if (FrameBuffer == NULL) {
+    return EFI_UNSUPPORTED;
+  }
+
+  Console = BuildGuidHob (&gLoaderFrameBufferConsoleGuid, sizeof (FRAME_BUFFER_CONSOLE));
+  InitFrameBufferConsole (FrameBuffer, Console,
+                          FrameBuffer->HorizontalResolution,
+                          FrameBuffer->VerticalResolution,
+                          0, 0);
+  PcdSet64(PcdFrameBufferConsoleBase, (UINT64)(UINTN)Console);
+
+  // If the console is configured to output messages to the serial port, also
+  // print out any messages that may be stored in output log buffer as the
+  // framebuffer console is not initialized until much later in the boot flow.
+  if ((PcdGet32 (PcdConsoleOutDeviceMask) & ConsoleOutSerialPort)
+      && (PcdGet32 (PcdDebugOutputDeviceMask) & DEBUG_OUTPUT_DEVICE_LOG_BUFFER)) {
+    LogBufHdr = (DEBUG_LOG_BUFFER_HEADER *) GetDebugLogBufferPtr ();
+    if (LogBufHdr != NULL) {
+      ConsoleWrite (LogBufHdr->Buffer, LogBufHdr->UsedLength - LogBufHdr->HeaderLength);
```

Comment:
To avoid duplicated message in serial port, how about this?
Unmask ConsoleOutSerialPort in PcdConsoleOutDeviceMask
ConsoleWrite
Mask again